corfudb-tools/src/main/java/org/corfudb/browser/CorfuOfflineBrowserEditor.java

```diff
@@ -0,0 +1,471 @@
+     * @param tableEntries SMR and Checkpoint updates.
+     */
+    private void processLogData(List<LogData> data,
+                                  CorfuTableDescriptor table,
+                                  CorfuRuntime runtimeSerializer,
+                                  List<LogEntryOrdering> tableEntries) {
+
+        for (int index = 0; index < data.size(); ++index) {
+
+            Object modifiedData = data.get(index).getPayload(runtimeSerializer);
+            List<SMREntry> smrUpdates = new ArrayList<>();
+            long snapshotAddress = 0;
+            if (modifiedData instanceof CheckpointEntry) {
+                snapshotAddress = Long.decode(((CheckpointEntry)modifiedData).
+                        getDict().
+                        get(CheckpointEntry.CheckpointDictKey.SNAPSHOT_ADDRESS));
+                MultiSMREntry smrEntries = ((CheckpointEntry) modifiedData).
+                        getSmrEntries(false, runtimeSerializer);
+                if (smrEntries != null) {
+                    smrUpdates = smrEntries.getUpdates();
+                }
+            } else if (modifiedData instanceof MultiObjectSMREntry) {
+                smrUpdates = ((MultiObjectSMREntry) modifiedData).getSMRUpdates(table.getStreamID());
+            }
+
+            for (SMREntry smrUpdate : smrUpdates) {
+                LogEntryOrdering entry = (modifiedData instanceof CheckpointEntry) ?
+                        new LogEntryOrdering(smrUpdate, snapshotAddress) :
+                        new LogEntryOrdering(smrUpdate, smrUpdate.getGlobalAddress());
+                tableEntries.add(entry);
+            }
+        }
+    }
+
+    /**
+     * Builds the given table.
+     * @param namespace
+     * @param tableName
+     * @return ConcurrentMap
+     */
+    public ConcurrentMap getTableData(String namespace, String tableName) {
+
+        CorfuTableDescriptor table = new CorfuTableDescriptor(namespace, tableName);
+        List<CorfuTableDescriptor> tableDsc = Arrays.asList(registryTableDsc, protobufDescTableDsc, table);
+
+        List<LogData> registryTableEntries = new ArrayList<>();
+        List<LogData> protobufDescriptorTableEntries = new ArrayList<>();
+        List<LogData> dataEntries = new ArrayList<>();
+        List<List<LogData>> tableData = Arrays.asList(registryTableEntries, protobufDescriptorTableEntries, dataEntries);
+
+        parseLogs(tableDsc, tableData);
+
+        CorfuRuntime serializer = CorfuRuntime.fromParameters(CorfuRuntime.CorfuRuntimeParameters.builder().build());
+        serializer.getSerializers().registerSerializer(DynamicProtobufSerializer.createProtobufSerializer());
+
+        List<LogEntryOrdering> registryEntries = new ArrayList<>();
+        ConcurrentMap registryTable = new ConcurrentHashMap<>();
+        processLogData(registryTableEntries, registryTableDsc, serializer, registryEntries);
+        addRecordsToConcurrentMap(registryEntries, registryTable);
```

Comment:
Let's consider the case where checkpointing has been failing repeatedly - this will result in the database file carrying several gigabytes of repeated updates to the same table. In that case won't this offline browser tool go out of memory when it attempts to keep all the log data entries in memory. Or am I missing something?

Reply:
Yes, it does. As discussed in the previous comment, we had thought of building an indexer for the lists and the concurrent maps to avoid OOM issues.

Reply:
The core usecase of this tool is in case of customer deployments where something bad has already happened. OOM is one of the very common "bad things" that happen. So we should probably not merge something that assumes everything is running fine.

Reply:
Got it. I've modified the logic to only cache the key-value pairs, instead of the transactions.